### PR TITLE
[JENKINS-58705] Update default working dir to /home/jenkins/agent

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-58705](https://issues.jenkins-ci.org/browse/JENKINS-58705)

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 1]
```diff
@@ -266,7 +266,7 @@ public void setRemoteFs(String remoteFs) {
 
     @Deprecated
     public String getRemoteFs() {
-        return getFirstContainer().map(ContainerTemplate::getWorkingDir).orElse(null);
+        return getFirstContainer().map(ContainerTemplate::getWorkingDir).orElse(ContainerTemplate.DEFAULT_WORKING_DIR);
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: If left null, not sure exactly why, but the workspaces still default to `/home/jenkins/workspace` and this breaks all the `container` calls.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 2]
```diff
@@ -290,7 +294,7 @@ public Pod build() {
                 // Running on OpenShift Enterprise, security concerns force use of arbitrary user ID
                 // As a result, container is running without a home set for user, resulting into using `/` for some tools,
                 // and `?` for java build tools. So we force HOME to a safe location.
-                env.put("HOME", workingDir);
+                env.put("HOME", DEFAULT_HOME);
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: For the lack of a better alternative...

**Comment by jglick** [comment]
failures in `PodTemplateBuilderTest`

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java (hunk 1 above)
Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: This method is deprecated; is it still used in production?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 3]
```diff
@@ -258,6 +260,8 @@ public Pod build() {
                 LOGGER.log(Level.INFO, "Computer is null for agent: {0}", slave.getNodeName());
             }
 
+            env.put("JENKINS_AGENT_WORKDIR", workingDir);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Is this necessary? I thought just setting the remote FS would suffice.

File: src/main/webapp/help/workingDir.html [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Path to the root of the workspace from the view point of this node, such as <em>/home/jenkins/agent</em>.<br/>
+This doesn't need to be absolute since the launcher establishes a consistent working directory, such as <em>./.jenkins-agent</em>.
```
Comment (jglick) [inline src/main/webapp/help/workingDir.html]: I am not sure what this second sentence is about. I think it is copied from some core help? Probably better to delete it; users should not rely on the CWD of the init process.

**Comment by jglick** [comment]
What this PR is missing is a new test which fails when `/home/jenkins` is used.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java (hunk 1 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Yes, unfortunately used by KubernetesSlave constructor. I have some needed refactoring in mind to make this better.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 3 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: Not strictly necessary, just thought it would be nicer

**Comment by Vlatombe** [comment]
> What this PR is missing is a new test which fails when /home/jenkins is used.

Not sure what you mean. I think pod templates using the old path should still work. It's just a new default after all.

**Comment by SteffenSeckler** [comment]
fyi: I got an error using the old /home/jenkins directory. Using /home/jenkins/agent in the pod templates worked, however. The error was:
```
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from ............(hidden).............
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1743)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
		at hudson.remoting.Channel.call(Channel.java:957)
		at hudson.FilePath.act(FilePath.java:1072)
		at hudson.FilePath.act(FilePath.java:1061)
		at hudson.FilePath.mkdirs(FilePath.java:1246)
		at hudson.plugins.git.GitSCM.createClient(GitSCM.java:821)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1169)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:124)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
		at java.lang.Thread.run(Thread.java:745)
java.nio.file.AccessDeniedException: /home/jenkins/workspace
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384)
	at java.nio.file.Files.createDirectory(Files.java:674)
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
	at java.nio.file.Files.createDirectories(Files.java:767)
	at hudson.FilePath.mkdirs(FilePath.java:3239)
	at hudson.FilePath.access$1300(FilePath.java:212)
	at hudson.FilePath$Mkdirs.invoke(FilePath.java:1254)
	at hudson.FilePath$Mkdirs.invoke(FilePath.java:1250)
	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3052)
	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:369)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:93)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

**Comment by Vlatombe** [comment]
@SteffenSeckler Would you mind filing a JIRA with your pod template definition?

**Comment by SteffenSeckler** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-58766

**Comment by stephenmuss** [comment]
This is a major breaking change and has caused all sorts of headaches for us. We've relied on `$HOME` for a fair bit of stuff and now it's not possible to write to that path.

This change feels like a sledgehammer approach to the problem it was trying to solve.

Unfortunately, AWS have made changes to EKS which caused issues in the older version of the plugin. So we're stuck with a plugin that doesn't work whichever version we try to use.

I expect this is going to impact a significant number of other users too.

**Comment by Vlatombe** [comment]
@stephenmuss Please open a JIRA, describe your use case, provide a reproducer for your problem. Complaining about a problem isn't going to solve it.

**Comment by ReignbowRoad** [comment]
@stephenmuss Can you elaborate on the changes AWS have made to EKS that have caused issues with the older version of KubernetesPlugin? Which version were/are you using? I was using version 1.17.3 of the plugin which was working a couple of weeks ago and now it won't provision any pods when I run my jobs in jenkins.

**Comment by stephenmuss** [comment]
@ReignbowRoad they released some patches for the Go net/http CVE. See here too https://issues.jenkins-ci.org/browse/JENKINS-59000?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&showAll=true

**Comment by marckhouzam** [comment]
This caused problems for us too.  We've worked around it but here is the info if it may help others.

In our case we run kubectl commands from our Jenkins workers.  So we would install our Kubernetes credentials in /home/jenkins/kubeconfig through the jnlp container and then have the other containers set a KUBECONFIG variable to /home/jenkins/kubeconfig.

With this change /home/jenkins/kubeconfig was not visible to the other containers and they could no longer talk to kubernetes.  So we had to change every Jenkinsfile that had hard-coded the use of /home/jenkins/kubeconfig to use a new /home/jenkins/agent/kubeconfig.

**Comment by andyfeller** [comment]
Gotta say this is fundamental the wrong way to implement this change as it should have been elective rather than prescribed as there are plenty of factors that hinge on the workspace location that will cause anyone running Jenkins with Kubernetes / OpenShift to have significant breakage.

I'm currently still trying to triage why setting the `workingDir` attribute of the declarative pipeline doesn't merge this in with the `yaml` provided but I'm stuck because this has become a blocking depending on `openshift-sync` for a significant race condition causing multiple builds being triggered.  tl;dr: this is more impactful than presumed

**Comment by chips-bo** [comment]
just hit this bug too and using version 1.26 of the plugin on jenkins v 2.222. All our jobs that run on /home/jenkins are now failing.

**Pull request merged**